dopamine/discrete_domains/run_experiment.py

```diff
@@ -607,37 +622,58 @@ def _run_experiment_loop(self):
     iterations is run.
     """
     threads = []
+    train_iterations = threading.Semaphore(self._max_simultaneous_iterations)
+    eval_iterations = threading.Semaphore(1)
     self._completed_iteration = self._start_iteration
+
     for iteration in range(self._start_iteration, self._num_iterations):
-      self._running_iterations.acquire()
-      threads.append(self._run_one_iteration(iteration))
+      if (iteration + 1) % self._eval_period == 0:
+        threads.append(self._run_one_iteration(
+            eval_iterations, iteration, True))
+      threads.append(self._run_one_iteration(
+          train_iterations, iteration, False))
+
     # Wait for all running iterations to complete.
     for thread in threads:
       thread.join()
 
   @threaded_method
-  def _run_one_iteration(self, iteration):
+  def _run_one_iteration(self, iteration, eval_mode):
     """Runs one iteration in separate thread, logs and checkpoints results.
 
     Same as parent Runner implementation except that summary statistics are
     directly logged instead of being returned.
 
     Args:
       iteration: int, current iteration number, used as a global_step for saving
-        Tensorboard summaries
+        Tensorboard summaries.
+      eval_mode: bool, whether this is an evaluation iteration.
```

Comment:
I think using an `enum` makes sense. It could also be the [TF estimator mode keys](https://www.tensorflow.org/api_docs/python/tf/estimator/ModeKeys).

The advantage that I see is that it would improve clarity for the developer. Is it also what you are thinking?

If that is the case, I would add it as a TODO, as it is independent from this CL and requires change across Dopamine.

Reply:
Nice.  Using ModeKeys sounds like like the winner to me.

I added a line to our refactor bullpen.  It works for me as a TODO, too.

Reply:
Added as a TODO.